search-backend-indexer/src/main/java/org/apache/maven/index/search/backend/indexer/internal/IndexerCoreSearchBackendImpl.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.maven.index.GroupedSearchResponse;
+import org.apache.maven.index.Indexer;
+import org.apache.maven.index.IteratorSearchRequest;
+import org.apache.maven.index.IteratorSearchResponse;
+import org.apache.maven.index.SearchType;
+import org.apache.maven.index.context.IndexingContext;
+import org.apache.maven.index.expr.SourcedSearchExpression;
+import org.apache.maven.index.grouping.GAGrouping;
+import org.apache.maven.index.search.MAVEN;
+import org.apache.maven.index.search.SEARCH;
+import org.apache.maven.index.search.SearchRequest;
+import org.apache.maven.index.search.backend.indexer.IndexerCoreSearchBackend;
+import org.apache.maven.index.search.backend.indexer.IndexerCoreSearchResponse;
+import org.apache.maven.index.search.request.Field;
+import org.apache.maven.index.search.request.Paging;
+import org.apache.maven.index.search.support.SearchBackendSupport;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * An engine to perform search trough single repository index (endpoint).
+ */
+public class IndexerCoreSearchBackendImpl extends SearchBackendSupport implements IndexerCoreSearchBackend
+{
+    private static final Map<Field, org.apache.maven.index.Field> FIELD_TRANSLATION;
+
+    static
+    {
+        HashMap<Field, org.apache.maven.index.Field> map = new HashMap<>();
+        map.put( MAVEN.GROUP_ID, org.apache.maven.index.MAVEN.GROUP_ID );
+        map.put( MAVEN.ARTIFACT_ID, org.apache.maven.index.MAVEN.ARTIFACT_ID );
+        map.put( MAVEN.VERSION, org.apache.maven.index.MAVEN.VERSION );
+        map.put( MAVEN.CLASSIFIER, org.apache.maven.index.MAVEN.CLASSIFIER );
+        map.put( MAVEN.PACKAGING, org.apache.maven.index.MAVEN.PACKAGING );
+        map.put( MAVEN.CLASS_NAME, org.apache.maven.index.MAVEN.CLASSNAMES );
+        map.put( MAVEN.FQ_CLASS_NAME, org.apache.maven.index.MAVEN.CLASSNAMES );
+        map.put( MAVEN.SHA1, org.apache.maven.index.MAVEN.SHA1 );
+        FIELD_TRANSLATION = Collections.unmodifiableMap( map );
+    }
+
+    private final Indexer indexer;
+
+    private final IndexingContext indexingContext;
+
+    /**
+     * Creates backend instance using provided indexer and context.
+     */
+    public IndexerCoreSearchBackendImpl( Indexer indexer, IndexingContext indexingContext )
+    {
+        super( indexingContext.getId(), indexingContext.getRepositoryId() );
+        this.indexer = requireNonNull( indexer );
+        this.indexingContext = indexingContext;
+    }
+
+    @NotNull
+    @Override
+    public IndexingContext getIndexingContext()
+    {
+        return indexingContext;
+    }
+
+    @NotNull
+    @Override
+    public IndexerCoreSearchResponse search( SearchRequest searchRequest ) throws IOException
+    {
+        // if GA present in query: doing flat, otherwise grouped search to mimic SMO
+        HashSet<Field> searchedFields = new HashSet<>();
```

Comment:
why is this typed and the upcoming collections not?

Reply:
WDYM?